### PR TITLE
Add test suite framework (fixes #160)

### LINKED ISSUE #160: Complete folder structure for test suite
Each file should have a docstring explaining it's intended purpose.

### LINKED ISSUE #173: Add a code coverage tool

### PR DESCRIPTION
**Pull request by JackMorganNZ**
# Proposed changes

- Fixes #160: Complete folder structure for test suite.
- ~Fixes #173: Add a code coverage tool.~
  - ~Coverage threshold is set to 20% for this pull request (currently at 23%) and we can increase this over time. Otherwise every Travis build would fail at the initially set 50%.~
- Adds a page for #178 but content is yet to be added.
- Added new `csu` script commands:
  - `csu style` runs the `flake8` and `pydocstyle` commands within the Docker image (so don't need virtual environment). These commands can still be run locally.
  - `csu test_backwards` is for running the test suite backwards. Always runs locally but only for PRs when on Travis.

## Issues to create after merge

- *Change scope of Docker images:* Currently the local and production Docker images only have the contents of the `csunplugged` file added to them, which caused issues when running the `flake8`, `coverage`, and `pydocstyle` commands as their configuration files were not copied as they are in the root directory. I moved them into the `csunplugged` folder for this branch and works fine. However in the long term we may want to run these or new tests on other folders, for example, testing the documentation builds correctly. Therefore the scope of the Docker images needs to be altered to include these files.

### CODE REVIEW

File: csunplugged/tests/topics/models/test_test_connected_generated_resource.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+from model_mommy import mommy
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_connected_generated_resource.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_curriculum_area.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+from tests.BaseTestWithDB import BaseTestWithDB
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_curriculum_area.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_curriculum_integration.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+from model_mommy import mommy
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_curriculum_integration.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_learning_outcome.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from tests.BaseTestWithDB import BaseTestWithDB
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_learning_outcome.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_lesson.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+from tests.BaseTestWithDB import BaseTestWithDB
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_lesson.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_programming_exercise.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+from tests.BaseTestWithDB import BaseTestWithDB
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_programming_exercise.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_programming_exercise_difficulty.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+from tests.BaseTestWithDB import BaseTestWithDB
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_programming_exercise_difficulty.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_programming_exercise_language.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+from tests.BaseTestWithDB import BaseTestWithDB
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_programming_exercise_language.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_programming_exercise_language_implementation.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+from tests.BaseTestWithDB import BaseTestWithDB
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_programming_exercise_language_implementation.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_topic.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+from tests.BaseTestWithDB import BaseTestWithDB
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_topic.py]: Whoops this file is still called `test_test_`

File: csunplugged/tests/topics/models/test_test_unit_plan.py [hunk 11]
```diff
@@ -0,0 +1,7 @@
+from tests.BaseTestWithDB import BaseTestWithDB
```
Comment (ravenmaster001) [inline csunplugged/tests/topics/models/test_test_unit_plan.py]: Whoops this file is still called `test_test_`

File: .travis.yml [hunk 12]
```diff
@@ -20,15 +20,17 @@ before_install:
...
-# Runs test suite
 script:
-  - flake8
-  - pydocstyle --verbose --count --explain
+  - ./csu test
+  - ./csu test_backwards
+  - ./csu style
+
+after_script:
+  - /docker_venv/bin/coveralls
```
Comment (ravenmaster001) [inline .travis.yml]: This is not working and hasn't been for the last 13 commits. It needs access to the file generated by the `coverage run` command.

See our coverage report here: https://coveralls.io/github/uccser/cs-unplugged

File: csu [hunk 13]
```diff
@@ -102,6 +105,30 @@ function shell {
...
+  docker-compose run django /docker_venv/bin/flake8
+  echo "Running Python docstring checker..."
+  docker-compose run django /docker_venv/bin/pydocstyle --count --explain
+}
+
+# Run test suite with coverage checking
+function test_suite {
+  echo "Running test suite..."
+  docker-compose run django /docker_venv/bin/coverage run --rcfile=.coveragerc ./manage.py test --pattern "test_*.py" -v 3
+  docker-compose run django /docker_venv/bin/coverage report
```
Comment (ravenmaster001) [inline csu]: I wasn't using this for reporting because of coveralls, do we want both?

File: .travis.yml [hunk 14]
```diff
@@ -20,15 +20,17 @@ before_install:
...
-# Runs test suite
 script:
-  - flake8
-  - pydocstyle --verbose --count --explain
+  - ./csu test
+  - ./csu test_backwards
+  - ./csu style
+
+after_script:
+  - docker-compose run django /docker_venv/bin/coveralls
```
Comment (ravenmaster001) [inline .travis.yml]: Still don't see this reflected on (see comment from review is probably outdated from this change):
https://coveralls.io/github/uccser/cs-unplugged

File: csu (hunk 13 above)
Comment (JackMorganNZ) [inline csu]: This just outputs the coverage graph using `coverage`, but I've yet to figure out how you uploaded that coverage data using `coveralls`, is there supposed to be a `.coveralls.yml` file?

**Pull request merged**